### ISSUE TITLE
Bug - wrong scan type for decimal with null values

### ISSUE THREAD

**Issue by scottlepp** [SAP/go-hdb#105]
When executing the following query

```
SELECT null/1024 from M_VOLUME_FILES
```

The [scan type returned](https://github.com/SAP/go-hdb/blob/main/driver/internal/protocol/result.go#L82) is "driver.Decimal".  And when using that type to scan the row it fails [here](https://github.com/SAP/go-hdb/blob/1b6d4955e5921d9e8c751c7fbe12403cd81a8f75/driver/decimal.go#L16):  

Since this query returns null values, the scan type should be "driver.NullDecimal"

**Comment by stfnmllr** [comment]
fixed with https://github.com/SAP/go-hdb/releases/tag/v1.0.3

**Comment by scottlepp** [comment]
@stfnmllr reopening as this fix caused a similar issue.  The scan type returned for the query below is now **driver.NullDecimal**.  Using that to scan the row results in a null value, but the value is not null.  Using **driver.Decimal** works.

```
SELECT SCHEMA_NAME, AVG(MEMORY_SIZE_IN_TOTAL) AS \"Size\" FROM M_CS_TABLES GROUP BY SCHEMA_NAME ORDER BY \"Size\" DESC"
```

**Comment by stfnmllr** [comment]
@scottlepp this works as intended and we are talking about two different things:

- Using driver.Decimal works.
This works because the result is most probably never NULL. Interesting here is that the HANA backend reports that the result column might be NULL

- sql: Scan error on column index 1, name "Size": invalid decimal value <nil>
The error is caused because currently go-hdb expects that 

```
type NullDecimal struct {
	Decimal *Decimal
	Valid   bool // Valid is true if Decimal is not NULL
}
```

the Decimal field of NullDecimal is instantiated before using. Please see https://github.com/SAP/go-hdb/blob/main/driver/datatype_test.go for reference. While your pull request https://github.com/SAP/go-hdb/pull/108 is addressing this, you would find the same behavior for NullLob and NullBytes whereas NullLob is the most tricky one because the Lob constructor does at least need an io.Writer for scanning.

In summary: the current implementation is consistent (for Null* fields containing reference fields the reference field needs to be instantiated before using).
I am going to double check if a compatible consistent solution for ALL Null types would be possible. Only if this is the case an go-hdb update is going to be provided.

**Comment by scottlepp** [comment]
@stfnmllr thanks that makes sense.  Unfortunately we can't use driver.Decimal as we rely on the scantype returned from the database to create the [scannable row](https://github.com/grafana/grafana-plugin-sdk-go/blob/main/data/sqlutil/scanrow.go#L38).  This method works with all drivers, so it seems to be a bug that this driver is returning the wrong scan type.

**Comment by stfnmllr** [comment]
@scottlepp yes, I see. The problem using reflect functions for instantiating go-hdb Null* objects is because the wrapped objects (Lob, Decimal) are pointers and need to be instantiated as well. Unfortunately I cannot change this as it is an incompatible change affecting all users of the driver (something which only can make it to a new (v2) version).
Just to mention it again, the scan type is 'derived' by the information the HANA backend is returning to the driver.
In addition the scan methods are now instantiating all nil references with default values. So most of the potential issues should be fixed with
https://github.com/SAP/go-hdb/releases/tag/v1.0.7
Looking forward to your feedback

**Comment by scottlepp** [comment]
@stfnmllr 1.0.7 is working.  Thanks.
